daemon/cmd/devices.go

```diff
@@ -0,0 +1,337 @@
+	if !hasDefaultRoute {
+		_, virtual := link.(*netlink.Veth)
+		if virtual {
+			log.WithField(logfields.Device, name).
+				Warn("Ignoring veth device as it has no default route")
+			return false
+		}
+
+	}
+	return true
+}
+
+type linkInfo struct {
+	hasDefaultRoute bool
+}
+
+// updateDevicesFromRoutes processes a batch of routes and updates the set of
+// devices. Returns true if devices changed.
+func (dm *DeviceManager) updateDevicesFromRoutes(l3DevOK bool, routes []netlink.Route) bool {
+	linkInfos := make(map[int]linkInfo)
+
+	// Collect all link indices mentioned in the route update batch
+	for _, route := range routes {
+		// Only consider devices that have global unicast routes,
+		// e.g. skip loopback, multicast and link local routes.
+		if route.Dst != nil && !route.Dst.IP.IsGlobalUnicast() {
+			continue
+		}
+		if route.Table == unix.RT_TABLE_LOCAL {
+			continue
+		}
+		linkInfo := linkInfos[route.LinkIndex]
+		linkInfo.hasDefaultRoute = linkInfo.hasDefaultRoute || route.Dst == nil
+		linkInfos[route.LinkIndex] = linkInfo
+	}
+
+	changed := false
+	for index, info := range linkInfos {
+		link, err := netlink.LinkByIndex(index)
+		if err != nil {
+			log.WithError(err).WithField(logfields.LinkIndex, index).
+				Warn("Failed to get link by index")
+			continue
+		}
+		name := link.Attrs().Name
+
+		// Skip devices we already know.
+		if _, exists := dm.devices[name]; exists {
+			continue
+		}
+
+		viable := dm.isViableDevice(l3DevOK, info.hasDefaultRoute, link)
+		if viable {
+			dm.devices[name] = struct{}{}
+			changed = true
+		}
+	}
+	return changed
+}
+
+// expandDevices expands all wildcard device names to concrete devices.
+// e.g. device "eth+" expands to "eth0,eth1" etc. Non-matching wildcards are ignored.
+func expandDevices() error {
+	allLinks, err := netlink.LinkList()
+	if err != nil {
+		return fmt.Errorf("Device wildcard expansion failed to fetch devices: %w", err)
+	}
+	expandedDevices := make(map[string]struct{})
+	for _, iface := range option.Config.Devices {
+		if strings.HasSuffix(iface, "+") {
+			prefix := strings.TrimRight(iface, "+")
+			for _, link := range allLinks {
+				attrs := link.Attrs()
+				if strings.HasPrefix(attrs.Name, prefix) {
+					expandedDevices[attrs.Name] = struct{}{}
+				}
+			}
+		} else {
+			expandedDevices[iface] = struct{}{}
+		}
+	}
+	if len(option.Config.Devices) > 0 && len(expandedDevices) == 0 {
+		// User defined devices, but expansion yielded no devices. Fail here to not
+		// surprise with auto-detection.
+		return fmt.Errorf("Device wildcard expansion failed to detect devices. Please verify --%s option.",
+			option.Devices)
+	}
+
+	option.Config.Devices = make([]string, 0, len(expandedDevices))
+	for dev := range expandedDevices {
+		option.Config.Devices = append(option.Config.Devices, dev)
+	}
+	sort.Strings(option.Config.Devices)
+	return nil
+}
+
+func areDevicesRequired() bool {
+	return len(option.Config.Devices) == 0 &&
```

Comment:
This part of the condition feels a bit unexpected to me given the function's name. Whether devices are preconfigured or not shouldn't impact whether they are required. I would just remove it.